File: Chapitre_4/exo4_2.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>
#define BUFFER_SIZE 256

int main()
{
    int tube_1[2],tube_2[2],valeurW,n;
    pid_t fils;
    FILE * f;
    char bufferRead[256], bufferWrite[256],chaine[255],valeurR[256];
    if(pipe(tube_1)!=0) exit(1); // Creation d'un tube_1 : pour envoyer et recuperer le mot
    if(pipe(tube_2)!=0) exit(1); // Creation d'un tube_1 : pour envoyer et recuperer le resultat
    fils=fork(); //Creation du processus fils
    if (fils == -1) exit(0); //Si il y a une erreur dans la cretion du processus fils
    if(fils==0) {
        printf("Entrez un mot : "); 
        scanf("%s",bufferWrite);
        close(tube_1[0]);
        write(tube_1[1],bufferWrite,BUFFER_SIZE); //Ecrire le mot dans le tube 
        close(tube_2[1]);
        read(tube_2[0],valeurR,BUFFER_SIZE); //Lire le resultat dans le tube
        n=atoi(valeurR);
        if(n==1) { //Affichage du résultat 
            printf("Le mot %s est dans le fichier mots.txt \n",bufferWrite);
        }
        else if(n==0) {
            printf("Le mot %s n'est pas dans le fichier mots.txt \n",bufferWrite);
        }
    }
    else {
        dup2(tube_1[0], STDIN_FILENO); //Lier la sortie tube[0] du tube à stdin
        close(tube_1[1]);
        scanf("%s",bufferRead); //Lire le mot dans le tube
        printf("\n");
        f=fopen("mots.txt","r"); //Ouvrir le fichier mots.txt
        if(f!=NULL) {
            fgets(chaine,255,f);
            if(strstr(chaine,bufferRead)!=NULL) valeurW=1; //Chercher le mot dans le fichier
            else valeurW=0;
            fclose(f);
        }
        dup2(tube_2[1], STDOUT_FILENO); //Lier l’entrée tube[1] du tube à stdout
        close(tube_2[0]);
        printf("\n%d\n",valeurW); //Ecrire le résultat dans le tube 
        wait(NULL);
    }
    return(0);
}

File: Chapitre_4/exo4_2.py
import os

r1 , w1=os.pipe() # Creation de r1,w1 : pour envoyer et recuperer le mot
r2 , w2 = os.pipe() #Creation de r2 ,w2 : pour envoyer et recuperer le resultat
fils=os.fork() #Creation du processus fils
if fils==-1 : #Si il y a une erreur dans la cretion du processus fils
    os._exit(0)
if fils==0:
    bufferWrite=input("Entrez un mot : ")
    os.close(r1)
    os.write(w1,bufferWrite) #Ecrire le mot dans le tube 
    os.close(w1)
    os.close(w2)
    message=os.read(r2,1024) #Lire le resultat dans le tube
    os.close(r2)
    valeurR=int(message.decode()) 
    if valeurR==1: #Affichage du résultat 
        print("Le mot {} est dans le fichier mots.txt \n".format(bufferWrite))
    elif valeurR==0:
        print("Le mot {} n'est pas dans le fichier mots.txt \n".format(bufferWrite))
else:
    os.close(w1)
    os.dup2(r1,0)
    os.close(r1)
    bufferRead=input("")
    file=open("mots.txt","r") #Ouvrir le fichier mots.txt
    chaine=file.readline()
    ret=chaine.find(bufferRead) #Chercher le mot dans le fichier
    if(ret>=0):
        valeurW=1
    else:
        valeurW=0
    file.close()
    os.close(r2)
    valeurW=str(valeurW) 
    os.dup2(w2,1) #Ecrire le résultat dans le tube 
    os.close(w2)
    print("{}".format(valeurW))
    os.wait()